core/network-libp2p/src/transport.rs

```diff
@@ -34,39 +36,81 @@ pub fn build_transport(
 	keypair: identity::Keypair,
 	wasm_external_transport: Option<wasm_ext::ExtTransport>
 ) -> (Boxed<(PeerId, StreamMuxerBox), io::Error>, Arc<bandwidth::BandwidthSinks>) {
+	// Build configuration objects for encryption mechanisms.
+	#[cfg(not(target_os = "unknown"))]
+	let noise_config = {
+		let noise_keypair = noise::Keypair::new().into_authentic(&keypair)
+			// For more information about this panic, see in "On the Importance of Checking
+			// Cryptographic Protocols for Faults" by Dan Boneh, Richard A. DeMillo,
+			// and Richard J. Lipton.
+			.expect("can only fail in case of a hardware bug; since this signing is performed only \
+				once and at initialization, we're taking the bet that the inconvenience of a very \
+				rare panic here is basically zero");
+		noise::NoiseConfig::ix(noise_keypair)
+	};
+	let secio_config = secio::SecioConfig::new(keypair);
+
+	// Build configuration objects for multiplexing mechanisms.
 	let mut mplex_config = mplex::MplexConfig::new();
 	mplex_config.max_buffer_len_behaviour(mplex::MaxBufferBehaviour::Block);
 	mplex_config.max_buffer_len(usize::MAX);
+	let yamux_config = yamux::Config::default();
 
+	// Build the base layer of the transport.
 	let transport = if let Some(t) = wasm_external_transport {
 		OptionalTransport::some(t)
 	} else {
 		OptionalTransport::none()
 	};
-
 	#[cfg(not(target_os = "unknown"))]
 	let transport = {
 		let desktop_trans = tcp::TcpConfig::new();
 		let desktop_trans = websocket::WsConfig::new(desktop_trans.clone())
 			.or_transport(desktop_trans);
 		transport.or_transport(dns::DnsConfig::new(desktop_trans))
 	};
-
 	let (transport, sinks) = bandwidth::BandwidthLogging::new(transport, Duration::from_secs(5));
 
-	// TODO: rework the transport creation (https://github.com/libp2p/rust-libp2p/issues/783)
-	let transport = transport
-		.with_upgrade(secio::SecioConfig::new(keypair))
-		.and_then(move |out, endpoint| {
-			let peer_id = out.remote_key.into_peer_id();
+	// Encryption
+
+	// For non-WASM, we support both secio and noise.
+	#[cfg(not(target_os = "unknown"))]
+	let transport = transport.and_then(move |stream, endpoint| {
+		let upgrade = core::upgrade::SelectUpgrade::new(noise_config, secio_config);
+		core::upgrade::apply(stream, upgrade, endpoint)
+			.and_then(|out| match out {
+				// We negotiated noise
+				EitherOutput::First((remote_id, out)) => {
+					let remote_key = match remote_id {
+						noise::RemoteIdentity::IdentityKey(key) => key,
+						_ => return Err(upgrade::UpgradeError::Apply(EitherError::A(noise::NoiseError::InvalidKey)))
+					};
+					Ok((EitherOutput::First(out), remote_key.into_peer_id()))
+				}
+				// We negotiated secio
+				EitherOutput::Second(out) =>
+					Ok((EitherOutput::Second(out.stream), out.remote_key.into_peer_id()))
+			})
+	});
+
+	// For WASM, we only support secio for now.
```

Comment:
```suggestion
	// For WASM, we only support secio for now.  This will be fixed once we support alternative cryptograpy libraries.
```

Reply:
Well, we also want to keep support for secio at the moment because the details of noise are blurry and could change.

Reply:
I don't want to imply that supporting secio is a bug per se.

Reply:
IPFS has no plans to deprecate secio, so rust-libp2p should keep it, but rust-libp2p could feature gate it.  We might not need secio in substrate's build for polkadot depending upon what the parachains networks do.

Reply:
@tomaka

> I don't want to imply that supporting secio is a bug per se.

Not supporting noise when compiling for wasm is a bug, though.